src/dos/program_more.cpp

```diff
@@ -0,0 +1,508 @@
+}
+
+bool MORE::ParseCommandLine()
+{
+	// Put all the parameters into vector
+	std::vector<std::string> params;
+	cmd->FillVector(params);
+
+	// Check if specified tabulation size
+	if (!params.empty()) {
+		const auto &param = params[0];
+		if ((starts_with("/t", param) || starts_with("/T", param)) &&
+		    (param.length() == 3) && (param.back() >= '1') &&
+		    (param.back() <= '9')) {
+			// FreeDOS extension - custom TAB size
+			tab_size = static_cast<uint8_t>(param.back() - '0');
+			params.erase(params.begin());
+		}
+	}
+
+	// Make sure no other switches are supplied
+	for (const auto &param : params)
+		if (starts_with("/", param)) {
+			WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH"), param.c_str());
+			return false;
+		}
+
+	// Create list of input files
+	return FindInputFiles(params);
+}
+
+bool MORE::FindInputFiles(const std::vector<std::string> &params)
+{
+	input_files.clear();
+	if (params.empty())
+		return true;
+
+	constexpr auto search_attr = UINT16_MAX & ~DOS_ATTR_DIRECTORY &
+	                             ~DOS_ATTR_VOLUME;
+
+	RealPt save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+
+	for (const auto &param : params) {
+		// Retrieve path to current file/pattern
+		char path[DOS_PATHLENGTH];
+		if (!DOS_Canonicalize(param.c_str(), path))
+			continue;
+		char *const end = strrchr(path, '\\') + 1;
+		assert(end);
+		*end = 0;
+
+		// Search for the first file from pattern
+		if (!DOS_FindFirst(param.c_str(),
+		                   static_cast<uint16_t>(search_attr))) {
+			LOG_WARNING("DOS: MORE.COM - no match for pattern '%s'",
+			            param.c_str());
+			continue;
+		}
+
+		while (!shutdown_requested) {
+			CALLBACK_Idle();
+
+			char name[DOS_NAMELENGTH_ASCII];
+			uint32_t size = 0;
+			uint16_t time = 0;
+			uint16_t date = 0;
+			uint8_t attr  = 0;
+
+			DOS_DTA dta(dos.dta());
+			dta.GetResult(name, size, date, time, attr);
+			assert(name);
+
+			input_files.emplace_back();
+			auto &entry = input_files.back();
+
+			if (attr & DOS_ATTR_DEVICE) {
+				entry.is_device = true;
+				entry.name      = std::string(name);
+			} else {
+				entry.is_device = false;
+				entry.name = std::string(path) + std::string(name);
+			}
+
+			if (!DOS_FindNext())
+				break;
+		}
+	}
+
+	dos.dta(save_dta);
+
+	if (!shutdown_requested && input_files.empty()) {
+		WriteOut(MSG_Get("SHELL_CMD_MORE_NO_FILE"));
+		WriteOut("\n");
+		return false;
+	}
+
+	return true;
+}
+
+std::string MORE::GetShortName(const std::string &file_name, const char *msg_id)
+{
+	assert(msg_id);
+
+	// The shortest name we should be able to display is:
+	// - 3 dots
+	// - 1 path separator
+	// - 8 characters of name
+	// - 1 dot
+	// - 3 characters of extension
+	// This gives 16 characters.
+	// We need to keep the last column free (reduces max length by 1).
+	// Format string contains '%s' (increases max length by 2).
+	constexpr size_t min = 16;
+	const auto max_len   = std::max(min, max_columns - std::strlen(MSG_Get(msg_id)) + 1);
```

Comment:
Thanks for these comments; there was a lot more going into this max-length determination than I originally thought! Helps a lot.